### PR TITLE
Make nessie-demos usable (again)

### PR DESCRIPTION
**Pull request by snazy**
Currently all Nessie Demos are unusable. This change makes those usable again, and updates a bunch of dependencies. The changes in detail are:

* Bump Python from 3.7 to 3.10. Python 3.11 does not work, because of some issues with not-well maintained dependencies - Flink (a version that's supported by Iceberg), pandas + numpy.
* Bump Nessie from 0.30.0 0.74.0. This required to have both Java 8 and Java 11: Hive requires Java 8, Nessie at least Java 11, Spark Java 11. Add safety checks to validate that both Java 8 and Java 11 are configured.
* Bump pynessie from 0.30.0 to 0.65.0.
* Bump Flink from 1.13.6 to 1.17.1
* Bump Hadoop from 2.10.1 to 2.10.2
* Bump Iceberg from 0.13.1 to 1.4.2
* Bump a bunch of dependencies in `requirements_dev.txt` + `requirements_lint.txt`
* Bump dependencies in `pre-commit-config.yaml`
* Remove Delta-Demo due to dropped support for Delta Lake.
* GH Actions
  * Introduce `concurrency` group
  * Cache downloaded tools (Hadoop, Hive, Spark)
  * Add a check that the `FROM` statement in `Dockerfile` starts with the right prefix. The `build_run_local_docker.sh` script temporarily updates the `Dockerfile` by updating the `FROM` statement to make it work locally. Committing the temporary change is wrong.
  * Bump some actions versions
* Made startup of Nessie safer by adding a check whether the required port 19120 is listening.
* Made startup of Hive-server safer by:
  * Adding a check whether the required port 10000 (Thrift) is listening.
  * Save the PID of the started Hive server, kill an already running Hive server to have a clean state for repeated test runs (tox).
* Some files received whitespace (especially the notebooks) and formatting changes due to updated linter dependencies.
* Some changes in `tox.ini` due to stricter rule for `passenv` and removed dependency-resolution flag.

### CODE REVIEW

File: docker/binder/start.hive [hunk 1]
```diff
@@ -38,21 +40,84 @@ fi
...
+mkdir -p $HIVE_WAREHOUSE_DIR
+
 # Initialize Hive's Derby database
 $HIVE_HOME/bin/schematool -dbType derby -initSchema
 echo "Finished initializing Derby database for Hive."
 
 # increase the Heap memory being used by Hive-MapReduce jobs
 export HADOOP_HEAPSIZE=1500
 
+# Use Java 8 for Hive :facepalm:
```
Comment (omarsmak) [inline docker/binder/start.hive:89]: that is really annoying

**Review by omarsmak — APPROVED** [review]
Looks fine to me. But to be honest there are many things that I forgot about 😓.

File: docker/binder/start.hive (hunk 1 above)
Comment (snazy) [reply docker/binder/start.hive:89]: After 5+ years of Java 11 and 6+ years after Java 9, yea.

**Comment by snazy** [comment]
I'll merge tomorrow and see how it looks on mybinder.org

**Comment by snazy** [comment]
> > I'll merge tomorrow and see how it looks on mybinder.org
> 
> you could test before you merge, I think you can specify the repo (fork?) and the ref in mybinder

Sadly not :(
```
aiting for build to start...
Picked Git content provider.
Cloning into '/tmp/repo2dockernhw_q3h4'...
HEAD is now at 5738950 Make nessie-demos usable (again)
Using DockerBuildPack builder
Step 1/7 : FROM ghcr.io/projectnessie/nessie-binder-demos:dd32c4413d91c22676121f62119bcc7f167e4752
manifest unknown
```

**Comment by snazy** [comment]
> even if the mybinder stuff wasnt reliable that way you could still easily see what nessie can do for you, so maybe worth bringing back.

Guess it would be worth to have that automated, possibly from a tox run?

**Pull request merged**